LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/UdpServer.cs

```diff
@@ -89,60 +109,89 @@ async Task RunUdpListener()
                         {
                             pullAckRemoteLoRaAggregatorPort = receivedResults.RemoteEndPoint.Port;
                         }
-                        SendAcknowledgementMessage(receivedResults,(int)PhysicalIdentifier.PULL_ACK, receivedResults.RemoteEndPoint);
+                        SendAcknowledgementMessage(receivedResults, (int)PhysicalIdentifier.PULL_ACK, receivedResults.RemoteEndPoint);
                         break;
                     //This is a PUSH_DATA (upstream message).
                     case PhysicalIdentifier.PUSH_DATA:
+
                         SendAcknowledgementMessage(receivedResults, (int)PhysicalIdentifier.PUSH_ACK, receivedResults.RemoteEndPoint);
 
                         // Message processing runs in the background
-                        MessageProcessor messageProcessor = new MessageProcessor(this.configuration, this.loraDeviceInfoManager);
 
-                        // do not wait for it to return
-                        _ = Task.Run(async () => {
-                            try
+#pragma warning disable CS4014
+                        Task.Run(async () =>
+                        {
+                            List<Rxpk> messageRxpks = Rxpk.CreateRxpk(receivedResults.Buffer);
+                            if (messageRxpks != null)
                             {
-                                var resultMessage = await messageProcessor.ProcessMessageAsync(receivedResults.Buffer);
-                                if (resultMessage != null && resultMessage.Length > 0)
+                                if (messageRxpks.Count == 1)
                                 {
-                                    var port = pullAckRemoteLoRaAggregatorPort;
-                                    if (port <= 0)
-                                        port = receivedResults.RemoteEndPoint.Port;
-                                    await this.UdpSendMessage(resultMessage, receivedResults.RemoteEndPoint.Address.ToString(), port);
+                                     await ProcessRxpkAsync(receivedResults.RemoteEndPoint.Address.ToString(), messageRxpks[0], startTimeProcessing);
+                                }
+                                else {
+                                    
+                                   for (int i = 0; i< messageRxpks.Count-1;i++)
```

Comment:
Even though I advised you to do this way we might need to think how to optimize thread pool usage.
Assuming messageRxpks.Length == 1 for 95% of the cases we should aim to create the following amount of tasks:

|Rxpks|Tasks created (current)|Tasks created (ideal)|
|-|-|-|
|0|1|1|
|1|2|1|
|N|N+2|N+1|

Reply:
I am not sure I follow you on this, let's check tomorrow together.

Reply:
I agree with Xico that if we can save a new thread/task for the common case is a good idea

Reply:
We evaluated the changes required and there non-trivial. We would suggest putting this as a work item for the message processor refactor sprint.

Reply:
Ok